File: src/courses/data/catalog/02-dsa/dsa-core/implementations/javascript/searching.js
/**
 * JavaScript Implementation: Searching (Linear & Binary)
 * Invariant: Linear O(N), Binary O(log N)
 */
export class SearchingLinearBinary {
  constructor() {
    this.items = [];
  }

  // TODO: Implement verified invariant
  insert(val) {
    this.items.push(val);
  }

  get size() {
    return this.items.length;
  }
}

File: src/courses/data/catalog/02-dsa/dsa-core/implementations/python/searching.py
"""
Python Implementation: Searching (Linear & Binary)
Invariant: Linear O(N), Binary O(log N)
"""
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class SearchingLinearBinary(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    # TODO: Implement core operational method
    def insert(self, value: T) -> None:
        self._items.append(value)

    def __len__(self) -> int:
        return len(self._items)
